### PR TITLE
Amm Executor

### PR DESCRIPTION
**Pull request by Bromel777**

### CODE REVIEW

File: amm-executor/src/Spectrum/Executor/OrdersExecutor/Service.hs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+run'
+  :: forall s m era. (IsStream s, MonadAsync m, MonadUnliftIO m)
+  => Logging m
+  -> Backlog m
+  -> Transactions m era
+  -> PoolResolver m
+  -> PoolActions
+  -> s m ()
+run' logging@Logging{..} backlog@Backlog{..} txs resolver poolActions =
+  S.repeatM tryAcquire & S.mapM (\case
```
Comment (GusevTimofey) [inline amm-executor/src/Spectrum/Executor/OrdersExecutor/Service.hs:70]: Should we repeat with some constant delay?

**Comment by yasha-black** [comment]
Task linked: [CU-2hcpm5n Implement AMM Executor](https://app.clickup.com/t/2hcpm5n)

File: amm-executor/src/Spectrum/Executor/Types.hs [hunk 2]
```diff
@@ -46,8 +46,8 @@ newtype OrderId = OrderId TxOutRef
 
 type Pool = OnChain Core.Pool
 
-poolId :: Pool -> PoolId
-poolId (OnChain _ Core.Pool{poolId}) = poolId
+extractPoolId :: Pool -> PoolId
+extractPoolId (OnChain _ Core.Pool{poolId}) = poolId
```
Comment (oskin1) [inline amm-executor/src/Spectrum/Executor/Types.hs:50]: Why?

File: amm-executor/src/Spectrum/Executor/OrdersExecutor/Service.hs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  -> PoolActions
+  -> Either OrderExecErr (TxCandidate, Predicted Core.Pool)
+runOrder (OnChain poolOut pool) (OnChain orderOut Core.AnyOrder{..}) PoolActions{..} =
+  case anyOrderAction of
+    DepositAction deposit -> runDeposit (Confirmed orderOut deposit) (poolOut, pool)
+    RedeemAction redeem   -> runRedeem (Confirmed orderOut redeem) (poolOut, pool)
+    SwapAction swap       -> runSwap (Confirmed orderOut swap) (poolOut, pool)
+
+attachLogging :: Monad m => Logging m -> OrderExecErr -> Order -> (a -> m Bool) -> (a -> m ())
+attachLogging Logging{..} err order ma input = ma input >>= (\opRes -> infoM (show err ++ " for " ++ show order ++ ". Op result:" ++ show opRes))
```
Comment (oskin1) [inline amm-executor/src/Spectrum/Executor/OrdersExecutor/Service.hs:116]: `logResult` is better. At first I was confused by attachLogging as it is used to attach logging algebra to a module.

File: amm-executor/src/Spectrum/Executor/OrdersExecutor/Service.hs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+  -> s m ()
+run' logging@Logging{..} backlog@Backlog{..} txs resolver poolActions =
+  S.repeatM tryAcquire & S.mapM (\case
+      Just order ->
+        infoM ("Going to execute order for pool" ++ show order) >> execute' logging backlog txs resolver poolActions order
+      Nothing    ->
+        pure ()
+    )
+
+execute' :: forall m era. (MonadUnliftIO m, MonadThrow m) => Logging m -> Backlog m -> Transactions m era -> PoolResolver m -> PoolActions -> Order -> m ()
```
Comment (oskin1) [inline amm-executor/src/Spectrum/Executor/OrdersExecutor/Service.hs:77]: Fmt. The list of args doesn't fit into one screen.

File: amm-executor/src/Spectrum/Executor/OrdersExecutor/Service.hs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+      dropErr       -> attachLogging logging dropErr order drop $ orderId order
+    )
+
+executeOrder' :: (Monad m, MonadThrow m) => Backlog m -> Transactions m era -> PoolResolver m -> PoolActions -> Order -> m ()
+executeOrder' Backlog{checkLater} Transactions{..} PoolResolver{..} poolActions order@(OnChain _ Core.AnyOrder{..}) = do
+  mPool <- resolvePool anyOrderPoolId
+  pool@(OnChain prevPoolOut Core.Pool{poolId})  <- throwMaybe (EmptyPool anyOrderPoolId) mPool
+  (txCandidate, Predicted _ predictedPool)      <- throwEither $ runOrder pool order poolActions
+  tx    <- finalizeTx txCandidate
+  pPool <- throwMaybe (PoolNotFoundInFinalTx poolId) (extractPoolTxOut pool tx)
```
Comment (oskin1) [inline amm-executor/src/Spectrum/Executor/OrdersExecutor/Service.hs:90]: `PoolNotFoundInFinalTx` doesn't sound like a business exception.

Comment (oskin1) [reply amm-executor/src/Spectrum/Executor/OrdersExecutor/Service.hs:90]: I mean this shouldn't happen at all

Comment (Bromel777) [reply amm-executor/src/Spectrum/Executor/OrdersExecutor/Service.hs:110]: on one way - yes, but on another extractPoolTxOut couldn't guarantee that tx contains poolOut, so on my opinion PoolNotFoundInFinalTx is normal error here

File: amm-executor/src/Spectrum/Executor/OrdersExecutor/Service.hs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+import qualified CardanoTx.Interop as Interop
+import Cardano.Api (Tx)
+import qualified RIO.List as List
+import Spectrum.Executor.PoolTracker.Data.Traced (Traced(Traced, prevTxOutRef, tracedState))
+import CardanoTx.ToPlutus (ToPlutus(toPlutus))
+import Spectrum.Executor.Data.OrderState (OrderInState(InProgressOrder, SuspendedOrder))
+import qualified Spectrum.Executor.Data.State as State
+import Ouroboros.Network.Subscription.PeerState ()
+import Prelude hiding (drop)
+import RIO.Time (UTCTime, getCurrentTime)
```
Comment (oskin1) [inline amm-executor/src/Spectrum/Executor/OrdersExecutor/Service.hs:36]: Please fmt this mess

**Review by oskin1 — APPROVED** [review]

**Pull request merged**